From f15cbf8fef2fb185a38f42db4b11c45e3ac8ccc9 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <grinberg@compulab.co.il>
Date: Thu, 20 Dec 2012 09:46:30 +0200
Subject: ARM: i.MX6: cm-fx6: fix SST SPI flash support

cm-fx6 uses SST SPI flash on bus 1 and chip select 0.
Adjust the board data so the flash is detected and used appropriately.

Signed-off-by: Igor Grinberg <grinberg@compulab.co.il>
---
 arch/arm/mach-mx6/board-cm-fx6.c |   37 ++++++++++++++++++++++---------------
 1 file changed, 22 insertions(+), 15 deletions(-)

diff --git a/arch/arm/mach-mx6/board-cm-fx6.c b/arch/arm/mach-mx6/board-cm-fx6.c
index 57b5e0f..e735639 100644
--- a/arch/arm/mach-mx6/board-cm-fx6.c
+++ b/arch/arm/mach-mx6/board-cm-fx6.c
@@ -300,43 +300,50 @@ static const struct spi_imx_master mx6_arm2_spi_data __initconst = {
 };
 
 #if defined(CONFIG_MTD_M25P80) || defined(CONFIG_MTD_M25P80_MODULE)
-static struct mtd_partition m25p32_partitions[] = {
+static struct mtd_partition cm_fx6_spi_flash_partitions[] = {
 	{
-		.name	= "bootloader",
+		.name	= "uboot",
 		.offset	= 0,
-		.size	= 0x00100000,
+		.size	= SZ_512K + SZ_256K,
 	}, {
-		.name	= "kernel",
+		.name	= "uboot environment",
+		.offset	= MTDPART_OFS_APPEND,
+		.size	= SZ_256K,
+	}, {
+		.name	= "reserved",
 		.offset	= MTDPART_OFS_APPEND,
 		.size	= MTDPART_SIZ_FULL,
 	},
 };
 
-static struct flash_platform_data m25p32_spi_flash_data = {
-	.name		= "m25p32",
-	.parts		= m25p32_partitions,
-	.nr_parts	= ARRAY_SIZE(m25p32_partitions),
-	.type		= "m25p32",
+/*
+ * The default flash on cm-fx6 is "sst25vf016b".
+ * It is JEDEC compliant, so we don't specify the .type feild below.
+ */
+static struct flash_platform_data cm_fx6_spi_flash_data = {
+	.name		= "spi_flash",
+	.parts		= cm_fx6_spi_flash_partitions,
+	.nr_parts	= ARRAY_SIZE(cm_fx6_spi_flash_partitions),
 };
 #endif
 
-static struct spi_board_info m25p32_spi0_board_info[] __initdata = {
-#if defined(CONFIG_MTD_M25P80)
+static struct spi_board_info cm_fx6_spi0_board_info[] = {
+#if defined(CONFIG_MTD_M25P80) || defined(CONFIG_MTD_M25P80_MODULE)
 	{
 	/* The modalias must be the same as spi device driver name */
 	.modalias	= "m25p80",
 	.max_speed_hz	= 20000000,
 	.bus_num	= 0,
-	.chip_select	= 1,
-	.platform_data	= &m25p32_spi_flash_data,
+	.chip_select	= 0,
+	.platform_data	= &cm_fx6_spi_flash_data,
 	},
 #endif
 };
 
 static void spi_device_init(void)
 {
-	spi_register_board_info(m25p32_spi0_board_info,
-				ARRAY_SIZE(m25p32_spi0_board_info));
+	spi_register_board_info(cm_fx6_spi0_board_info,
+				ARRAY_SIZE(cm_fx6_spi0_board_info));
 }
 
 static void ddc_dvi_init(void)
-- 
1.7.9.5

